feat(AWS Lambda): Add `platform` option for container images

zanemountcastle opened the pull request:
Closes: #10098

This pull request addresses an issue where containerized functions built with Docker on non-x86 architectures (e.g. new ARM-based Apple M1 Macs) fail at runtime with the following error:

```
Launch error: fork/exec /lambda-entrypoint.sh: exec format error Entrypoint: [/lambda-entrypoint.sh] Cmd: [handler.handler] WorkingDir: [/var/task]
```

The fix for this to add `--platform=linux/amd64` as an argument into the `docker build` command. I added `platform` as an option when defining the image as follows to do this:

```
service: service-name
provider:
  name: aws
  ecr:
    images:
      baseimage:
        path: ./path/to/context
        file: Dockerfile.dev
        platform: linux/amd64
```

I verified that this fix works in practice and resolves the runtime error when building on an M1 Mac for AWS Lambda.

pgrzesik commented:
Hello @zanemountcastle - thanks a lot for submitting the PR. I'm currently busy with another assignments, but I will try to test it out and review it by the end of the week. Thanks for your patience :bow:

pgrzesik commented on docs/providers/aws/guide/functions.md:
```diff
@@ -206,7 +206,7 @@ Alternatively lambda environment can be configured through docker images. Image
 Serverless will create an ECR repository for your image, but it currently does not manage updates to it. An ECR repository is created only for new services or the first time that a function configured with an `image` is deployed. In service configuration, you can configure the ECR repository to scan for CVEs via the `provider.ecr.scanOnPush` property, which is `false` by default. (See [documentat ...
 
-In service configuration, images can be configured via `provider.ecr.images`. To define an image that will be built locally, you need to specify `path` property, which should point to valid docker context directory. Optionally, you can also set `file` to specify Dockerfile that should be used when building an image. It is also possible to define images that already exist in AWS ECR repository. In ...
+In service configuration, images can be configured via `provider.ecr.images`. To define an image that will be built locally, you need to specify `path` property, which should point to valid docker context directory. Optionally, you can also set `file` to specify Dockerfile that should be used when building an image. It is also possible to define images that already exist in AWS ECR repository. In ...
```
I think we should explicitly state that you should specify `linux/amd64` only when you're targeting Lambda that uses this architecture. It might be a different case when someone is using arm64-based Lambdas, which should work just fine (I was testing it previously and it was working perfectly). Additionally, we might mention the same - if you're on `x86_64` and you want to target `arm64` Lambda runtime, you might need to specify different `platform` setting

pgrzesik commented on lib/plugins/aws/provider.js:
```diff
@@ -2170,13 +2183,20 @@ Object.defineProperties(
               'ECR_IMAGE_BOTH_URI_AND_CACHEFROM_DEFINED_ERROR'
             );
           }
+          if (imageDefinedInProvider.uri && imageDefinedInProvider.platform) {
```
It would be great to also cover that case with an unit test.

pgrzesik reviewed (changes requested):
Thanks a lot @zanemountcastle 🙇 It looks really good, I only have minor remarks regarding documentation & tests and we should be good to go 🎉

zanemountcastle commented:
I updated the documentation to make the use of the `platform` option clearer and added a test for the case when `uri` and `platform` are both specified.

I noticed that tests for the cases where `uri` is specified as well as `buildArgs` and `cacheFrom` were not covered, so I also added those since they're very similar to the `platform` test.

@pgrzesik, let me know if my changes (especially to the documentation) resolve your concerns. I tried to make it clear in the documentation that this is an optional argument, but it's pretty much mandatory for ARM-based machines since Lambda by default runs on `x86`-based AWS Graviton2 CPUs: https://docs.aws.amazon.com/lambda/latest/dg/foundation-arch.html

Especially for organizations that use machines with multiple architectures (e.g. some Intel-based Mac/Linux machines and some newer M1-based Macs), this argument is pretty important to avoid runtime errors from containers build on some computers but not others. I imagine this will be more and more common as more people transition to Apple M1.

mnapoli commented on docs/providers/aws/guide/functions.md:
```diff
@@ -206,7 +206,15 @@ Alternatively lambda environment can be configured through docker images. Image
-In service configuration, images can be configured via `provider.ecr.images`. To define an image that will be built locally, you need to specify `path` property, which should point to valid docker context directory. Optionally, you can also set `file` to specify Dockerfile that should be used when building an image. It is also possible to define images that already exist in AWS ECR repository. In ...
+In service configuration, images can be configured via `provider.ecr.images`. To define an image that will be built locally, you need to specify `path` property, which should point to valid docker context directory. Optionally, you can also set `file` to specify Dockerfile that should be used when building an image. It is also possible to define images that already exist in AWS ECR repository. In ...
+
+Additionally, you can define arguments that will be passed to the `docker build` command:
```
```suggestion
Additionally, you can define arguments that will be passed to the `docker build` command via the following properties:
```

Just to make it clear that those are YAML properties (not `serverless` CLI arguments)

pgrzesik commented on docs/providers/aws/guide/functions.md:
```diff
@@ -206,7 +206,15 @@ Alternatively lambda environment can be configured through docker images. Image
+
+- `buildArgs`: With the `buildArgs` property, you can define arguments that will be passed to `docker build` command with `--build-arg` flag. They might be later referenced via `ARG` within your `Dockerfile`. (See [Documentation](https://docs.docker.com/engine/reference/builder/#arg))
+- `cacheFrom`: The `cacheFrom` property can be used to specify which images to use as a source for layer caching in the `docker build` command with `--cache-from` flag. (See [Documentation](https://docs.docker.com/engine/reference/builder/#usage))
+- `platform`: The `platform` property can be used to specify the architecture target in the `docker build` command with the `--platform` flag. If not specified, Docker will build for your computer's architecture by default. AWS Lambda typically uses `x86` infrastructure unless otherwise specified in the Lambda's runtime settings. In order to avoid runtime errors when building on an ARM-based mach ...
```
`x86 infrastructure` -> should be changed to architecture instead of infrastructure I believe

pgrzesik replied:
`Apple M1 MAc` -> `Apple M1 Mac` - small casing issue

pgrzesik replied:
`x86`-based Lambas -> `x86`-based Lambdas - small typo

zanemountcastle replied:
Good catches. Not sure how I missed those. Fixed in the latest commit.

pgrzesik reviewed (changes requested):
Thanks a lot @zanemountcastle - it looks great :+1: I only have small suggestions to docs (minor typos) and we're good to go :rocket:

zanemountcastle commented:
@pgrzesik Typos are fixed.

pgrzesik reviewed (approved):
Awesome job @zanemountcastle :clap: Thanks a lot for your contribution. Let's wait for @mnapoli's re-review and we're good to merge.

mnapoli reviewed (approved):
👍️

kevincua-umbocv commented:
Is platform possible for non container images/ecs?

I'm trying to perform a `serverless deploy --verbose -s $STAGE` for arm64 using an x86 MacBook without ecs. The problem is that the package keeps performing a docker run without a --platform flag which I would need for the python packages to be built on arm64.

kevincua-umbocv commented:
@pgrzesik Well, I use sls deploy and it creates a docker to test everything and deploy. I'm not really sure how to explain since I'm not that proficient with using serverless yet but basically I do not have ecr in my serverless.yml:

```
  ecr:
    images:
      baseimage:
        path: ./path/to/context
        platform: linux/arm64
```

Nevermind tho I was able to deploy from x86 pc a lambda with arm64 by including this in my serverless.yml and using CEMU. I dont recommend it tho because it's severely slow on a 4-core cpu macbook.

```
platform: linux/arm64
```

pgrzesik commented:
Thanks for answering @kevincua-umbocv 

> I use sls deploy and it creates a docker to test everything and deploy. 

Are you using some plugins that use docker containers underneath? Like `serverless-python-requirements`?

kevincua-umbocv commented:
Yes I am.

On Mon, 18 Jul 2022, 6:42 pm Piotr Grzesik, ***@***.***>
wrote:

> Thanks for answering @kevincua-umbocv <https://github.com/kevincua-umbocv>
>
> I use sls deploy and it creates a docker to test everything and deploy.
>
> Are you using some plugins that use docker containers underneath? Like
> serverless-python-requirements?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/10237#issuecomment-1187042539>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AY4AO7VEPXQGRC2A63TWIM3VUUYJDANCNFSM5H3EVVHQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pgrzesik commented:
So in that case it would be plugin-specific config. If I'm not mistaked, you could set it this way: https://github.com/serverless/serverless-python-requirements#extra-docker-arguments

kevincua-umbocv commented:
Cool, thanks! Will try this

On Mon, 18 Jul 2022, 6:59 pm Piotr Grzesik, ***@***.***>
wrote:

> So in that case it would be plugin-specific config. If I'm not mistaken,
> you could set it this way:
> https://github.com/serverless/serverless-python-requirements#extra-docker-arguments
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/10237#issuecomment-1187071907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AY4AO7TEVI2IO2IVVWBH2WTVUU2JHANCNFSM5H3EVVHQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
